### ISSUE TITLE
Dot matching has unexpected behavior.

### ISSUE THREAD

**Issue by mmachenry** [PuzzleTechHub/nutrimatic#5]
I tried this query "..ab.." and it matched "label" among many other things, some of which are rather odd to match that query. The preceding two dots on the regexp requires that there be two characters preceding the "ab" but in "label" there is only one character. A similar thing happens with underscore as well.

**Comment by egnor** [comment]
Hmm, I suspect this has to do with the treatment of spaces. I'm unable to repro with underscore -- \_\_ab\_\_ only matches six letter words. (Dots match spaces, underscores don't.) Can you show me an underscore based query with surprising results?

**Comment by Glench** [comment]
This is still a surprising result:

https://nutrimatic.org/?q=_%2B+is+the+new+black&go=Go

Compare with:

https://nutrimatic.org/?q=______+is+the+new+black&go=Go

**Comment by mmachenry** [comment]
Hm... maybe the underscore doesn't have this. I thought I saw something funny on there but I don't see it now. The dot behavior is strange though. You're right it probably does have to do with space behavior.

**Comment by egnor** [comment]
@Glench, you mean the "Computation limit reached" for the first one? That's unfortunately very common when you have a query which is open ended at the beginning and constrained at the end. Any little bit of constraint helps. For example, putting quotes around the whole thing ("_+ is the new black") gets you the expected answer. Otherwise you could have any phrase at the start, and that's a big search space, and it's just doing a best-first left-to-right search. (Using a beam search might work better in some cases, but I've never gotten around to trying that.)

@mmachenry, totally agreed the dot behavior is wrong. I'll see what I can do about that. I probably never noticed over all these years because I habitually use underscores (when puzzle solving/creating, I hardly ever want "alphanumeric or space"; puzzles tend to enumerate printable characters, not spaces).

**Comment by Glench** [comment]
Ah, that makes sense, but is a little strange to have to explain to people. Oh well!

**Comment by egnor** [comment]
Okay, the treatment of spaces is subtle!

I actually _added_ support for matching leading spaces a few years ago (https://github.com/egnor/nutrimatic/commit/6a3f0954503c948e3fcd8351bc713c5d336a876b). "Match spaces at the beginning of the expression", the commit message says. I wish I had included a motivating example for why I wanted this!

It makes a certain kind of sense; if you think about grepping a big Wikipedia corpus, there are plenty of instances of " label" in Wikipedia. But the results are still inarguably confusing. (Also, it's weird that we _didn't_ allow _trailing_ spaces to be matched, even though the same logic would apply.) Possibly _printing_ the spaces in the result in some visible way, so you'd see a visible space before "label" in the output, could help make it less weird.

For now I've reverted the space logic back to before that previous change. If I find an example where I really want leading spaces, I'll make sure to comment exactly why in the code. :-) Marking this bug as closed; ..ab.. does a more intuitive seeming thing now. (You'll still get the dots matching internal spaces; hopefully that isn't too surprising.)

I did add code to strip whitespace from around the form input, just in case some extra spaces sneak in; literal leading and trailing spaces are unlikely to be desired.
